### PR TITLE
Add NumPy 2.0 support

### PR DESCRIPTION
**Pull request by aganders3**
This introduces numpy 2.0 support. napari has started looking into support, and ran into a few small issues using vispy testing against it. See the [napari zulip thread](https://napari.zulipchat.com/#narrow/stream/212875-general/topic/handling.20the.20numpy.202.2E0.20release/near/381330412) for some discussion and relevant links.

The process for adding support in this PR was:
* Run the ruff numpy 2.0 support check and fix:
```
ruff check vispy --select NPY201
ruff check vispy --select NPY201 --fix
```
* Run tests, fix issues and warnings
* Update for changes to behavior of the `copy` kwarg in `np.array`
    * Add `vispy.util.np_copy_if_needed` value (`False` for numpy < 1.28, `None` for numpy >= 2.0)
    * Search for `copy=False` used in `no.array` and remove it
    * Search for `copy=copy` used in `np.array`, replace with `copy=vispy.util.np_copy_if_needed`
    * See references for more information
        * [NumPy 2.0 Migration Guide](https://numpy.org/devdocs/numpy_2_0_migration_guide.html#adapting-to-changes-in-the-copy-keyword)
        * [SciPy PR doing the same](https://github.com/scipy/scipy/pull/20172)
        * [napari PR doing the same](https://github.com/napari/napari/pull/6932)
* Update build configuration for Cython code
    * Previously, it was important to link against the oldest compatible numpy version to ensure
      forward compatibility. This was accomplished by depending on
      [oldest-supported-numpy](https://pypi.org/project/oldest-supported-numpy/).
    * Now, you can just depend on `numpy>=2.0.0rc1`, and the build will be compatible with both 1.x
      and 2.x.
    * This is updated in the `[build-system]` section of `pyproject.toml`
    * See [Depending on NumPy](https://numpy.org/devdocs/dev/depending_on_numpy.html#numpy-2-0-specific-advice) for more information

I think this is ready for review, but I'm going to try creating a wheel and using to run napari tests with numpy 2.0 as well. I'll mark it non-draft after I do that.

### CODE REVIEW

**Comment by aganders3** [comment]
Ah, wheel builds are failing, which I did not test. It looks like maybe the NumPy 2.0-compatible wheels will need to be Python 3.9+.

**Comment by djhoese** [comment]
Do we build 3.8? I'm ok only doing 3.9+

**Comment by aganders3** [comment]
Yes, as of 0.14.2 there are 3.8 wheels. The first thing I noticed though was the sdist was apparently being built with 3.7 😮.

**Comment by aganders3** [comment]
Running tests in napari does reveal some deprecation warnings coming through vispy:
```
======================================================================================================== warnings summary ========================================================================================================
napari/layers/shapes/_tests/test_shapes_mouse_bindings.py: 49 warnings
  /Users/aandersoniii/src/napari/numpy2-venv/lib/python3.11/site-packages/vispy/geometry/triangulation.py:679: DeprecationWarning: Arrays of 2-dimensional vectors are deprecated. Use arrays of 3-dimensional vectors instead. (deprecated in NumPy 2.0)
    return np.cross(B-A, C-B) > 0

napari/layers/shapes/_tests/test_shapes_mouse_bindings.py: 18 warnings
  /Users/aandersoniii/src/napari/numpy2-venv/lib/python3.11/site-packages/vispy/geometry/triangulation.py:742: DeprecationWarning: Arrays of 2-dimensional vectors are deprecated. Use arrays of 3-dimensional vectors instead. (deprecated in NumPy 2.0)
    c = np.cross(v1, v2)  # positive if v1 is CW from v2
```

There is some discussion here about what should be used instead:
https://github.com/numpy/numpy/issues/26620

**Comment by brisvag** [comment]
Looks like we should wait for `cross2d` to be exposed; let's check back in a bit!

**Comment by aganders3** [comment]
I'm not sure we'd want to use it, as that would make it 2.0-only?

**Comment by brisvag** [comment]
True... I guess we should use their example like you did in the napari PR:

```py
cross2d = arr1[..., 0] * arr2[..., 1] - arr1[..., 1] * arr2[..., 0]
```

A bit ugly to lose the explicit "cross", but oh well

**Comment by aganders3** [comment]
Yeah, it's unfortunate but I understand why it's changing. My concern is if any of these functions can take 2D or 3D input, but the behavior is quite different so I doubt it.

I guess we could make our own `cross2d` to drop in.

**Comment by jni** [comment]
> I guess we could make our own cross2d to drop in.

👆 this is the right way I think. Abstract away the unreadability and make clear the intent of the code.

File: vispy/visuals/text/_sdf_cpu.pyx [hunk 1]
```diff
@@ -7,6 +7,8 @@ cimport numpy as np
 from libc.math cimport sqrt
 cimport cython
 
+np.import_array()
```
Comment (brisvag) [inline vispy/visuals/text/_sdf_cpu.pyx:10]: This is also from ruff fixes?

**Review by brisvag — APPROVED** [review]
Great, I like the new function, much more readable.

File: vispy/visuals/text/_sdf_cpu.pyx (hunk 1 above)
Comment (aganders3) [reply vispy/visuals/text/_sdf_cpu.pyx:10]: Actually no, this is from a warning. I thought it was related to numpy 2.0, but now that I look maybe not.

See this example in the [Cython docs](https://cython.readthedocs.io/en/latest/src/tutorial/numpy.html#adding-types):

```
# "cimport" is used to import special compile-time information
# about the numpy module (this is stored in a file numpy.pxd which is
# distributed with Numpy).
# Here we've used the name "cnp" to make it easier to understand what
# comes from the cimported module and what comes from the imported module,
# however you can use the same name for both if you wish.
cimport numpy as cnp

# It's necessary to call "import_array" if you use any part of the
# numpy PyArray_* API. From Cython 3, accessing attributes like
# ".shape" on a typed Numpy array use this API. Therefore we recommend
# always calling "import_array" whenever you "cimport numpy"
cnp.import_array()
```

File: pyproject.toml [hunk 2]
```diff
@@ -1,8 +1,9 @@
 [build-system]
 requires = [
     "wheel",
-    "oldest-supported-numpy",
     "setuptools>=42",
+    # see https://numpy.org/devdocs/dev/depending_on_numpy.html#numpy-2-0-specific-advice
+    "numpy>=2.0.0rc2",
```
[the remark this answers is not in the document]
Comment (aganders3) [reply pyproject.toml:6]: I don't think we were getting that for free before either, as I believe this should only affect the build configuration. It seems like they're recommending new tests added to test against the minimum supported version of numpy. Do you think we should add that here?

[the remark this answers is not in the document]
Comment (aganders3) [reply pyproject.toml:6]: What's happening in this branch now is that we're building against numpy 2.0.0rc2 and testing against the latest released version (1.26.4).

[the remark this answers is not in the document]
Comment (brisvag) [reply pyproject.toml:6]: Got it... No need to add them here then I guess :) We can do in a follow up!

[the remark this answers is not in the document]
Comment (aganders3) [reply pyproject.toml:6]: I think we'll need to do some testing to figure out what the minimum supported version of numpy should even be. Right now it's left unspecified in `setup.py`.

**Comment by jni** [comment]
Since tests are passing, we should probably merge and release this sooner rather than later since NumPy 2 is going to be out in a couple of days... 💣 😅

**Comment by brisvag** [comment]
Yep, seems good to me. Let's go!

**Comment by brisvag** [comment]
Released `0.14.3` so we can join the big "update to support numpy 2 party" that all projects are currently participating it :P

**Comment by aganders3** [comment]
Thanks! Fingers are crossed 😬
